README.md
# Virtual Repos

![preview](https://img.shields.io/badge/-preview-orange)

<!--[![Publish Extension](https://github.com/carlocardella/vscode-VirtualRepos/actions/workflows/PublishExtension.yml/badge.svg)](https://github.com/carlocardella/vscode-VirtualRepos/actions/workflows/PublishExtension.yml)-->
![Visual Studio Marketplace Version](https://img.shields.io/visual-studio-marketplace/v/carlocardella.vscode-virtualRepos)
![Visual Studio Marketplace Installs](https://img.shields.io/visual-studio-marketplace/i/carlocardella.vscode-virtualRepos)
![Visual Studio Marketplace Downloads](https://img.shields.io/visual-studio-marketplace/d/carlocardella.vscode-virtualRepos)
![Visual Studio Marketplace Rating](https://img.shields.io/visual-studio-marketplace/r/carlocardella.vscode-virtualRepos)
[![GitHub issues](https://img.shields.io/github/issues/carlocardella/vscode-VirtualRepos.svg)](https://github.com/carlocardella/vscode-VirtualRepos/issues)
[![GitHub license](https://img.shields.io/github/license/carlocardella/vscode-VirtualRepos.svg)](https://github.com/carlocardella/vscode-VirtualRepos/blob/master/LICENSE.md)
[![Twitter](https://img.shields.io/twitter/url/https/github.com/carlocardella/vscode-VirtualRepos.svg?style=social)](https://twitter.com/intent/tweet?text=Wow:&url=https%3A%2F%2Fgithub.com%2Fcarlocardella%2Fvscode-VirtualRepos)
<!-- [![Open in Visual Studio Code](https://open.vscode.dev/badges/open-in-vscode.svg)](https://open.vscode.dev/carlocardella/vscode-texttoolbox) -->

[Download for VS Code](https://marketplace.visualstudio.com/items?itemName=CarloCardella.vscode-virtualrepos)

<!-- [Download for VS Codium](https://open-vsx.org/extension/carlocardella/vscode-texttoolbox) -->

Virtual Repos is a Visual Studio Code extension that allows to open and edit a remote repository (e.g. on GitHub) without cloning, committing or pushing your changes. It all happens automatically.

The extension is still missing lots of features I want to add (as time permits) and you can expect bugs (but hopefully nothing destructive), anyway this is a `preview` extension and you can expect bugs here and there. Please report bugs or issues and ask for features you would like to see. Check [Changelog](CHANGELOG.md) for the latest status, what's planned and what has already been released.

## Getting started

Install the extension from the VSCode Marketplace.

## Repository management

### Open a repo

Use the `Open repository` command to open an existing repo from GitHub, you can choose from three options:

1. `Open repository`: open any repository you have access to, enter the name as `owner/repoName`

   *Note: If you just enter the repo name, the extension assumes you own it, this is the same as using `Open my repository`*

2. `Open my repository`: open a repository from a list of repos you own
3. `Open starred repository`: open one of your Starred repos

The repository will load automatically (of course, make sure you are connected to the Internet), you can then browse it, open, edit, add, delete files as if they where on your local file system even without cloning the repo.

*open repository*

![open repository](https://user-images.githubusercontent.com/5784415/192892207-46f5418e-5696-4373-ae80-71cb160e8e25.gif)

*open my repository*

![open my repository](https://user-images.githubusercontent.com/5784415/192892464-bee3d23f-5688-4dfd-a343-c844ae39e135.gif)

#### Sync repositories across devices

You can sync your open repositories across multiple devices by enabling [Settings Sync](https://code.visualstudio.com/docs/editor/settings-sync) in Visual Studio Code.

*Node: you may need to Refresh the Virtual Repos view to see the latest repos added or removed from another machine.*

### Create new repo

You can create a new repository (public or private).

![create private repo](https://user-images.githubusercontent.com/5784415/192894098-2cb95397-6696-467a-ab9c-6ca272f460b0.gif)

### Fork repository

If you opened a repository you do not own, you can now fork it from the context menu. Once forked, the repository is automatically added to the Virtual Repositories list.

### Star/Unstar repository

If you opened a repository you do not own, you can now star or unstar it. The list of starred repositories is automatically refreshed from GitHub every hour.

### Follow/Unfollow a user

If you opened a repository you do not own, you can follow or unfollow its owner. The list of followed users is automatically refreshed from GitHub every hour.

## Automatic commits

Changes are committed automatically after the file is saved. The commit message is `VirtualRepos: update file <filePath>`.

## Easily identify forked and private repositories

Public repositories are represented with a `$(repo)` icon from from your [IconTheme](https://code.visualstudio.com/api/references/icons-in-labels) or the built-in `codicon`, while private repositories use a `$(lock)` icon.
Forked repos use a `$(repo-forked)` icon instead, to clearly identity an open repository as a fork. This also helps distinguish between the `upstream` repository (represented with the `repo` icon) and my fork (represented with the `repo-fork` icon):

<!--![image](https://user-images.githubusercontent.com/5784415/207214066-1e78d96b-b209-4e7c-a263-4a225db1ce92.png)-->
![image](https://user-images.githubusercontent.com/5784415/210154421-d8eda221-d096-42fe-98e3-fc5f37773f58.png)

### Get repository info

You can get basic details about a repository by hovering the mouse on it:

![image](https://user-images.githubusercontent.com/5784415/210154457-25a31281-bf1e-430c-8f73-efddbfa15dce.png)

## Repo owner's avatar

You can identity repositories using their GitHub owner's avatar, just switch `VirtualRepos.UseRepoOwnerAvatar` to `true` (default is `false`) to replace the standard repo icon with the repo owner's GitHub avatar:

![UserRepoOwnerAvatar](https://user-images.githubusercontent.com/5784415/194788262-4eb6ad81-f924-4a42-a642-eec4c10a60b0.gif)

## Tracing

The extension uses the built-in Output Channel to enable local tracing to a channel called "Virtual Repositories". This data is local only, it does not leave your machine.

![image](https://user-images.githubusercontent.com/5784415/227406009-1775db2a-7e0f-4f01-9c89-a7714bb7e2da.png)

You can control the tracing level or disable, for all Output Channels or for specific extensions, using the built-in command `Developer: Set Log Level...`:

![image](https://user-images.githubusercontent.com/5784415/227406186-61ab711e-2a59-4376-b759-0efdc364ec1b.png)

## GitHub authentication scopes

The extension uses the following [GitHub authentication scopes](https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps)

- `user:email`: needed to authenticate to GitHub
- `user:follow`: needed to follow/unfollow a user on GitHub
- `repo`: needed to create and update repositories (and their content) for the authenticated user
- `delete_repo`: needed to delete repositories

## My other extensions

<!-- * [Virtual Repos](https://github.com/carlocardella/vscode-VirtualRepos): Virtual Repos is a Visual Studio Code extension that allows to open and edit a remote repository (e.g. on GitHub) without cloning, committing or pushing your changes. It all happens automatically -->
- [Virtual Gists](https://github.com/carlocardella/vscode-VirtualGists): Virtual Gists is a Visual Studio Code extension that allows to open and edit a remote gist (e.g. on GitHub) without cloning, committing or pushing your changes. It all happens automatically
- [Virtual Git](https://github.com/carlocardella/vscode-VirtualGit): VSCode extension path with my extensions to work with virtual repositories and gists based on a virtual file system
- [Text Toolbox](https://github.com/carlocardella/vscode-TextToolbox): Collection of tools for text manipulation, filtering, sorting etc...
- [File System Toolbox](https://github.com/carlocardella/vscode-FileSystemToolbox): VSCode extension to work with the file system, path auto-complete on any file type
- [Changelog Manager](https://github.com/carlocardella/vscode-ChangelogManager): VSCode extension, helps to build a changelog for your project, either in markdown or plain text files. The changelog format follows Keep a changelog
- [Hogwarts colors for Visual Studio Code](https://github.com/carlocardella/hogwarts-colors-for-vscode): Visual Studio theme colors inspired by Harry Potter, Hogwarts and Hogwarts Houses colors and banners

## Acknowledgements

Virtual Repositories is freely inspired by these fine extensions:

- [GistPad](https://marketplace.visualstudio.com/items?itemName=vsls-contrib.gistfs)
- [WikiLens](https://marketplace.visualstudio.com/items?itemName=lostintangent.wikilens)
- [GitDoc](https://marketplace.visualstudio.com/items?itemName=vsls-contrib.gitdoc)
- [Dendron](https://marketplace.visualstudio.com/items?itemName=dendron.dendron)
- [Foam](https://marketplace.visualstudio.com/items?itemName=foam.foam-vscode)


CHANGELOG.md
# Change Log

All notable changes to the "vscode-VirtualRepos" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

See the [list of open enhancements on GitHub](https://github.com/carlocardella/vscode-VirtualRepos/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Aenhancement)
## [0.1.2] - 2023-04-07

### Fixed

* [#48 Cached Starred Repos and My Repos are not refreshed](https://github.com/carlocardella/vscode-VirtualRepos/issues/48)

## [0.1.1] - 2023-04-03

### Changed

* `Open Repository...` allows to open multiple repos at once, when selecting from `Open my repository` and `Open starred repository`

### Fixed

* Fixed a couple of bugs around global storage handling and refreshing the treeview

## [0.1.0] - 2023-03-23

### Changed

* Updated minimum VSCode required version to 1.66 so support the updated GitHub authentication flow
* For VSCode version 1.74 or higher, use the built-in [Log Output Channel](https://code.visualstudio.com/updates/v1_74#_log-output-channel)
* Removed custom tracing logic for VSCode versions older than 1.74
## [0.0.31] - 2023-01-10

### Fixed

* [#45 Sort menu duplicate commands](https://github.com/carlocardella/vscode-VirtualRepos/issues/45)

## [0.0.30] - 2023-01-07

### Changed

* Reordered inline commands for folder and repositories
* Update repository tooltip
* Tweak tracing messages
* Improve TreeView refresh, show refresh progress bar more consistently

## [0.0.29] - 2023-01-03

### Changed

* The sort commands menu uses "toggle" items: the currently selected sert type and direction are identified with a checkmark

### Fixed

* Minor bug fixes and code refactoring

## [0.0.28] - 2022-12-31

### Added

* Toggle GitHub repository visibility (public or private)

### Fixed

* [#42 Error when creating new file under a folder (/folder/newfile.md)](https://github.com/carlocardella/vscode-VirtualRepos/issues/42)
* Various other smaller fixes

### Changed

* Private repositories are represented by a lock icon

## [0.0.27] - 2022-12-30

### Added

* Sort repositories by name, stars, forks, watchers, creation time, update time
* `Copy upstream url` (available for forked repos only)
* `Show upstream` (available for forked repos only)

### Changed

* Renamed `Show on remote` to `Show remote` to be consistent with the new `Show upstream`

See the [list of open enhancements on GitHub](https://github.com/carlocardella/vscode-VirtualRepos/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Aenhancement)

## [0.0.26] - 2022-12-24

### Added

* [#37 Follow//unfollow a user on GitHub](https://github.com/carlocardella/vscode-VirtualRepos/issues/37)
* The extension now requires `user:follow` authentication context on GitHub, this is needed to be able to follow/unfollow a user. You will need to re-authorize the extension in VSCode after update

## [0.0.25] - 2022-12-22

### Fixed

* [#38 Fix star/unstar icon in treeview at extension startup](https://github.com/carlocardella/vscode-VirtualRepos/issues/38)

## [0.0.24] - 2022-12-22

### Added

* Star and unstar an open repository
  * Starred repositories are refreshed from GitHub every hour
  * You can also use the `Refresh starred repositories from GitHub` command to force the refresh as needed

## [0.0.23] - 2022-12-13

### Changed

* Forked repos use a "forked" icon rather than the generic "repo" icon
* Upload multiple files in onw shot: [#1](https://github.com/carlocardella/vscode-VirtualRepos/issues/1)

## [0.0.22] - 2022-12-12

### Added

* Select and delete multiple files in a repo with one click

### Changed

* Improved new file experience: now the newly created file is automatically opened in the editor

## [0.0.21] - 2022-12-05

### Added

* Rename file/folder command in context menu
* Delete folder and all its content, command in context menu
  * _Known issue_: The `Delete folder` currently works only with folders in the repository root, this is being tracked in [#29](https://github.com/carlocardella/vscode-VirtualRepos/issues/29)

### Fixed

* Restored `Clone` and `Fork` context menu commands, removed by mistake in the previous release

## [0.0.20] - 2022-11-26

### Added

* `Fork repository`: if you opened a repository you do not own, you can now fork it from the context menu. Once forked, the repository is automatically added to the Virtual Repositories list
* View repository owner profile on GitHub

### Fixed

* Fixed write/delete actions in context menu, show then only when the user has write permission on the repository

## [0.0.19] - 2022-11-18

### Added

* Added `Copy remote url` in context menu
* Added `Show on remote` in context menu

### Changed

* Updated node modules

## [0.0.18] - 2022-11-15

### Fixed

* Fixed [#13](https://github.com/carlocardella/vscode-VirtualRepos/issues/13): Wrong refresh rage with VirtualRepos.PullInterval

## [0.0.17] - 2022-11-09

### Added

* Clone repository

### Changed

* Updated context menu commands
* Code refactoring and various bug fixes and improvements

## [0.0.16] - 2022-10-11

### Fixed

* Fix for [Files do not open on github.dev](https://github.com/carlocardella/vscode-VirtualRepos/issues/7)

src/GitHub/commands.ts
import { commands, env, Uri, window, workspace, ProgressLocation } from "vscode";
import { RepoFileSystemProvider, REPO_SCHEME } from "../FileSystem/fileSystem";
import { ContentNode, RepoNode } from "../Tree/nodes";
import {
    getGitHubRepoContent,
    newGitHubRepository,
    deleteGitHubRepository,
    getGitHubReposForAuthenticatedUser,
    getStarredGitHubRepositories,
    getGitHubUser,
    forkGitHubRepository,
    FileMode,
    TypeMode,
    createGitHubTree,
    createGitHubCommit,
    updateGitHubRef,
    unstarGitHubRepository,
    starGitHubRepository,
    unfollowGitHubUser,
    followGitHubUser,
    getGutHubFollowedUsers,
    updateGitHubRepository,
} from "./api";
import { TContent, TTreeRename, TRepo, TUser } from "./types";
import { credentials, extensionContext, output, repoFileSystemProvider, repoProvider, store } from "../extension";
import { byteArrayToString, charCodeAt, getFileNameFromUri, removeLeadingSlash, stringToByteArray } from "../utils";
import { StorageKeys } from "./constants";

/**
 * Returns the binary content of a file in the repository.
 * @date 8/26/2022 - 6:51:03 PM
 *
 * @export
 * @async
 * @param {TRepo} repo The repository to get the file from.
 * @param {string} filePath The path to the file in the repository.
 * @returns {Promise<Uint8Array>}
 */
export async function getRepoFileContent(repo: RepoNode, file: TContent): Promise<Uint8Array> {
    let data: any;
    if (!file!.content) {
        data = await getGitHubRepoContent(repo.owner, repo.name, file!.path);
        file!.content = data;
    } else {
        data = file!.content;
    }

    return new Uint8Array(Buffer.from(data.content, "base64").toString("latin1").split("").map(charCodeAt));
}

/**
 * Returns repository owner and name from <owner>/<repo> string
 *
 * @export
 * @param {string} repo The repository string
 * @returns {[string, string]}
 */
export function getRepoDetails(repo: string): [string, string] {
    const parts = repo.split("/");
    return [parts[0], parts[1]];
}

/**
 * QuickPick items for the user to select which repository to open
 *
 * @enum {number}
 */
enum QuickPickItems {
    repoName = "$(rocket) Open repository",
    userRepos = "$(account) Open my repository",
    starredRepos = "$(star) Open starred repository",
}

/**
 * Ask the user to choose or enter a repository to open: <owner>/<repo>
 *
 * @export
 * @async
 * @returns {(Promise<string | undefined>)}
 */
export async function pickRepository() {
    return await new Promise((resolve, reject) => {
        let pick: string | string[] | undefined;

        let quickPick = window.createQuickPick();
        quickPick.onDidHide(() => quickPick.dispose());
        quickPick.placeholder = "Select or type the repository you would like to open";
        quickPick.canSelectMany = false;
        quickPick.matchOnDescription = true;

        quickPick.show();

        quickPick.items = [{ label: QuickPickItems.repoName }, { label: QuickPickItems.userRepos }, { label: QuickPickItems.starredRepos }];

        quickPick.onDidAccept(async () => {
            if (pick === QuickPickItems.repoName || (Array.isArray(pick) && pick.includes(QuickPickItems.repoName))) {
                let repo = await window.showInputBox({
                    ignoreFocusOut: true,
                    placeHolder: "owner/repo",
                    title: "Enter the repository to open, e.g. 'owner/repo'",
                });
                if (repo) {
                    quickPick.items = [{ label: `${repo}` }];
                }
                quickPick.hide();
                resolve([repo]);
            } else if (pick === QuickPickItems.userRepos || (Array.isArray(pick) && pick.includes(QuickPickItems.userRepos))) {
                quickPick.busy = true;
                quickPick.placeholder = "Enter the repository to open, e.g. 'owner/repo'";
                const repos = extensionContext.globalState.get(StorageKeys.myRepos) as TRepo[];
                quickPick.busy = false;
                quickPick.canSelectMany = true;
                quickPick.items = repos!.map((repo) => ({ label: `${repo}` }));
                quickPick.show();
            } else if (pick === QuickPickItems.starredRepos || (Array.isArray(pick) && pick.includes(QuickPickItems.starredRepos))) {
                quickPick.busy = true;
                quickPick.placeholder = "Enter the repository to open, e.g. 'owner/repo'";
                const starredRepos = extensionContext.globalState.get(StorageKeys.starredRepos) as string[];
                quickPick.busy = false;
                quickPick.canSelectMany = true;
                quickPick.items = starredRepos.map((repo) => ({ label: repo }));
                quickPick.show();
            } else {
                output?.debug(`onDidAccept: ${pick}`);
                quickPick.hide();
                resolve(pick);
            }
        });

        quickPick.onDidChangeSelection(async (selection) => {
            pick = selection.map((item) => item.label);
            output?.debug(`onDidChangeSelection: ${pick}`);
        });
    });
}

/**
 * Create a new file in a GitHub repository
 *
 * @export
 * @async
 * @param {ContentNode} e The TreeItem containing the file to create
 * @returns {*}
 */
export async function addFile(e: ContentNode | RepoNode): Promise<Uri | undefined> {
    const newFileName = await window.showInputBox({ ignoreFocusOut: true, placeHolder: "path/filename", title: "Enter the filename (optional path)" });
    if (!newFileName) {
        return;
    }

    const [repoOwner, repoName, path] = RepoFileSystemProvider.getFileInfo(e.uri)!;
    const content = "";

    const newFileUri = Uri.parse(`${REPO_SCHEME}://${repoOwner}/${repoName}/${path}/${newFileName}`); // @ugly: if the file is created under root, path is empty and the output is '//'
    // prettier-ignore
    await repoFileSystemProvider.writeFile(
        newFileUri, stringToByteArray(content), {
        create: true,
        overwrite: true,
    });

    return Promise.resolve(newFileUri);
}

/**
 * Delete the file or folder from the repository
 *
 * @export
 * @async
 * @param {ContentNode} nodes TreeView nodes to delete
 * @returns {*}
 */
export async function deleteFile(nodes: ContentNode[]) {
    let confirm: "Yes" | "No" | undefined = undefined;
    let message: string;
    nodes.length === 1
        ? (message = `Are you sure you want to delete '${nodes[0].label}'?`)
        : (message = `Are you sure you want to delete ${nodes.length} files?`);

    confirm = await window.showWarningMessage(message, { modal: true }, "Yes", "No");
    if (confirm !== "Yes") {
        return;
    }

    let tree = nodes.map((node) => {
        return {
            path: node.path,
            mode: FileMode.file,
            type: TypeMode.blob,
            sha: null,
        } as TTreeRename;
    });

    const repo = store.repos.find((repo) => repo!.full_name === `${nodes[0].owner}/${nodes[0].repo.name}`)!;

    let newTree = await createGitHubTree(repo, tree);
    let newCommit = await createGitHubCommit(repo, `Delete ${tree.map((t) => t.path)}`, newTree!.sha!, [repo.tree?.sha!]);
    let updatedRef = await updateGitHubRef(repo, `heads/${repo.repo.default_branch}`, newCommit!.sha);

    if (updatedRef) {
        output?.info(`Deleted ${tree.map((t) => t.path)}`);
        repoProvider.refresh(repo);
    }
}

/**
 * Upload files from the local disk to GitHub
 *
 * @export
 * @async
 * @param {(ContentNode | RepoNode)} destination The destination to upload the files to
 * @returns {Promise<void>}
 */
export async function uploadFiles(destination: ContentNode | RepoNode): Promise<void> {
    const files = await window.showOpenDialog({ canSelectFiles: true, canSelectFolders: false, canSelectMany: true, title: "Select the files to upload" });
    if (!files) {
        return Promise.reject();
    }

    await window.withProgress({ title: "Uploading files...", location: ProgressLocation.Notification }, async () => {
        let tree = await Promise.all(
            files.map(async (file) => {
                return {
                    path: removeLeadingSlash(destination.path + "/" + getFileNameFromUri(file)),
                    mode: FileMode.file,
                    type: TypeMode.blob,
                    content: byteArrayToString(await workspace.fs.readFile(file)),
                } as TTreeRename;
            })
        );

        const repo = store.repos.find((repo) => repo!.full_name === `${destination.repo.owner.login}/${destination.repo.name}`)!;

        let newTree = await createGitHubTree(repo, tree);
        let newCommit = await createGitHubCommit(repo, `Upload ${tree.map((t) => t.path)}`, newTree!.sha!, [repo.tree?.sha!]);
        let updatedRef = await updateGitHubRef(repo, `heads/${repo.repo.default_branch}`, newCommit!.sha);

        if (updatedRef) {
            output?.info(`Uploaded ${tree.map((t) => t.path)}`);
            repoProvider.refresh();
        }
    });

    return Promise.reject();
}

/**
 * Create a new repository
 * @date 9/26/2022 - 10:06:19 AM
 *
 * @export
 * @async
 * @param {boolean} isPrivate Whether the repository should be private or not
 * @returns {Promise<void>}
 */
export async function newRepository(isPrivate: boolean): Promise<void> {
    const newRepo = await window.showInputBox({ ignoreFocusOut: true, placeHolder: "repo name", title: "Enter the repository name" });
    if (!newRepo) {
        return Promise.reject();
    }

    let [owner, repoName] = ["", ""];
    if (newRepo.indexOf("/") === -1) {
        owner = credentials.authenticatedUser.login;
        repoName = newRepo;
    } else {
        [owner, repoName] = newRepo.split("/");
    }

    // create the repository
    const githubRepo = await newGitHubRepository(owner, repoName, isPrivate);

    if (githubRepo) {
        //add the new repository to the tree view
        await store.addRepo(extensionContext, `${githubRepo.owner.login}/${githubRepo.name}`);
    }
}

/**
 * Delete a repository
 * @date 9/26/2022 - 10:05:56 AM
 *
 * @export
 * @async
 * @param {RepoNode} repo The repository to delete
 * @returns {Promise<void>}
 */
export async function deleteRepository(repo: RepoNode): Promise<void> {
    const confirm = await window.showWarningMessage(`Are you sure you want to delete '${repo.name}'?`, { modal: true }, "Yes", "No");
    if (confirm !== "Yes") {
        return Promise.reject();
    }

    const deleted = await deleteGitHubRepository(repo.repo);
    if (deleted) {
        await store.removeRepo(extensionContext, `${repo.repo.owner.login}/${repo.name}`);
    }
}

/**
 * Clone the selected repository
 *
 * @export
 * @async
 * @param {RepoNode} repo The repository to clone
 * @returns {*}
 */
export async function cloneRepository(repo: RepoNode) {
    output?.info(`Cloning ${repo.repo.clone_url}`);
    commands.executeCommand("git.clone", repo.repo.clone_url);
}

/**
 * Copy the repository or file URL to the clipboard
 *
 * @export
 * @param {(RepoNode | ContentNode)} node The node to copy the URL for
 */
export function copyRemoteUrl(node: RepoNode | ContentNode) {
    let url: string | undefined;

    if (node instanceof RepoNode) {
        url = node.repo.html_url;
    }
    if (node instanceof ContentNode) {
        url = node.nodeContent?.html_url;
    }

    if (url) {
        env.clipboard.writeText(url);
        output?.info(`Copied ${url} to the clipboard`);
    }
}

/**
 * Shows the selected repository or file on remote
 *
 * @export
 * @param {(RepoNode | ContentNode)} node The node to show on remote
 */
export function showRemote(node: RepoNode | ContentNode) {
    let url: string | undefined;

    if (node instanceof RepoNode) {
        url = node.repo.html_url;
    }
    if (node instanceof ContentNode) {
        url = node.nodeContent?.html_url;
    }

    if (url) {
        env.openExternal(Uri.parse(url));
        output?.info(`Show ${url} on remote`);
    }
}

/**
 * View the repository owner's profile on GitHub
 *
 * @export
 * @async
 * @param {string} username The username of the owner
 * @returns {*}
 */
export async function viewRepoOwnerProfileOnGitHub(username: string) {
    const user = await getGitHubUser(username);
    if (user) {
        env.openExternal(Uri.parse(user.html_url));
    }
}

/**
 * Fork a repository into the authenticated user's account
 *
 * @export
 * @async
 * @param {RepoNode} repo The repository to fork
 * @returns {unknown}
 */
export async function forkRepository(repo: RepoNode) {
    const forkedRepo = await forkGitHubRepository(repo.repo);
    if (forkedRepo) {
        await store.addRepo(extensionContext, `${forkedRepo.owner.login}/${forkedRepo.name}`);
        output?.info(`Forked ${repo.name} to ${forkedRepo.owner.login}/${forkedRepo.name}`);
        return Promise.resolve();
    }

    return Promise.reject();
}

/**
 * Rename a repo file
 *
 * @export
 * @async
 * @param {ContentNode} file The file to rename
 * @returns {*}
 */
export async function renameFile(file: ContentNode) {
    const newFileName = await window.showInputBox({
        ignoreFocusOut: true,
        placeHolder: "new file name",
        title: "Enter the new file name",
        value: file.label as string,
    });
    if (!newFileName) {
        return;
    }

    const newUri = RepoFileSystemProvider.getFileUri(file.repo, newFileName);
    const oldUri = file.uri;
    output?.info(`Rename "${file.path}" to "${newFileName}"`);
    await repoFileSystemProvider.rename(oldUri, newUri, { overwrite: false });
}

/**
 * Delete a folder from a repository
 *
 * @export
 * @async
 * @param {ContentNode} folder The folder to delete
 * @returns {*}
 */
export async function deleteFolder(folder: ContentNode) {
    const confirm = await window.showWarningMessage(`Are you sure you want to delete "${folder.path}"?`, { modal: true }, "Yes", "No");
    if (confirm !== "Yes") {
        return;
    }

    output?.info(`Delete "${folder.path}"`);
    await repoFileSystemProvider.deleteDirectory(folder.uri);
}

/**
 * Star or unstar a repository
 *
 * @export
 * @async
 * @param {RepoNode} repo The repository to star or unstar
 * @returns {*}
 */
export async function toggleRepoStar(repo: RepoNode) {
    let starredRepos = await getOrRefreshStarredRepos();
    if (repo.isStarred) {
        await unstarGitHubRepository(repo);
        starredRepos = starredRepos?.filter((r) => r !== repo.full_name);
        repo.isStarred = false;
        output?.info(`Unstarred ${repo.full_name}`);
    } else {
        await starGitHubRepository(repo);
        starredRepos?.push(repo.full_name);
        repo.isStarred = true;
        output?.info(`Starred ${repo.full_name}`);
    }

    await getOrRefreshStarredRepos(starredRepos);
}

/**
 * Return or refresh the list of starred repositories
 *
 * @export
 * @async
 * @param {?(string[] | TRepo[])} [starredRepoNames] The list of starred repositories
 * @param {?boolean} [forceRefreshFromGitHub] Force refresh from GitHub
 * @returns {(Promise<string[] | undefined>)}
 */
export async function getOrRefreshStarredRepos(starredRepoNames?: string[] | TRepo[], forceRefreshFromGitHub?: boolean): Promise<string[] | undefined> {
    if (!starredRepoNames) {
        starredRepoNames = extensionContext.globalState.get<string[]>(StorageKeys.starredRepos, []);
    }

    if (starredRepoNames?.length === 0 || forceRefreshFromGitHub) {
        output?.info("Fetching starred repositories from GitHub");
        starredRepoNames = await getStarredGitHubRepositories();
        starredRepoNames = starredRepoNames.map((repo) => `${repo.owner.login}/${repo.name}`);
        extensionContext.globalState.update(StorageKeys.starredRepos, starredRepoNames);
    } else {
        starredRepoNames = starredRepoNames as string[];
    }

    extensionContext.globalState.update(StorageKeys.starredRepos, starredRepoNames);
    commands.executeCommand("setContext", StorageKeys.starredRepos, starredRepoNames);
    return Promise.resolve(starredRepoNames);
}

export async function getOrRefreshAuthenticatedUserRepos(repoNames?: string[] | TRepo[], forceRefreshFromGitHub?: boolean): Promise<string[] | undefined> {
    if (!repoNames) {
        repoNames = extensionContext.globalState.get<string[]>(StorageKeys.myRepos, []);
    }

    if (repoNames?.length === 0 || forceRefreshFromGitHub) {
        output?.info("Fetching my repositories from GitHub");
        repoNames = await getGitHubReposForAuthenticatedUser();
        repoNames = repoNames.map((repo) => `${repo.owner.login}/${repo.name}`);
        extensionContext.globalState.update(StorageKeys.myRepos, repoNames);
    } else {
        repoNames = repoNames as string[];
    }

    extensionContext.globalState.update(StorageKeys.myRepos, repoNames);
    commands.executeCommand("setContext", StorageKeys.myRepos, repoNames);
    return Promise.resolve(repoNames);
}

/**
 * Follow or unfollow a user on GitHub
 *
 * @export
 * @async
 * @param {string} user The user to follow or unfollow
 * @returns {*}
 */
export async function toggleFollowUser(user: string) {
    let followingUsers = await getOrRefreshFollowedUsers();
    if (followingUsers) {
        let isFollowedUser = followingUsers?.includes(user);

        if (isFollowedUser) {
            await unfollowGitHubUser(user);
            followingUsers = followingUsers?.filter((u) => u !== user);
            output?.info(`Unfollowed ${user}`);
        } else {
            await followGitHubUser(user);
            followingUsers?.push(user);
            output?.info(`Followed ${user}`);
        }

        await getOrRefreshFollowedUsers(followingUsers);
    }
}

/**
 * Return or refresh the list of followed users
 *
 * @export
 * @async
 * @param {?(string[] | TUser[])} [followedUsers] The list of followed users
 * @param {?boolean} [forceRefreshFromGitHub] Force refresh from GitHub
 * @returns {(Promise<string[] | undefined>)}
 */
export async function getOrRefreshFollowedUsers(followedUsers?: string[] | TUser[], forceRefreshFromGitHub?: boolean): Promise<string[] | undefined> {
    if (!followedUsers) {
        followedUsers = extensionContext.globalState.get<string[]>("followedUsers", []);
    }

    if (followedUsers?.length === 0 || forceRefreshFromGitHub) {
        output?.info("Fetching followed users from GitHub");
        followedUsers = await getGutHubFollowedUsers();
        followedUsers = followedUsers!.map((user) => user.login);
        extensionContext.globalState.update("followedUsers", followedUsers);
    } else {
        followedUsers = followedUsers as string[];
    }

    extensionContext.globalState.update("followedUsers", followedUsers);
    commands.executeCommand("setContext", "followedUsers", followedUsers);
    return Promise.resolve(followedUsers);
}

/**
 * Copy the upstream URL to the clipboard
 *
 * @export
 * @async
 * @param {RepoNode} repo The repository to copy the upstream URL from
 * @returns {*}
 */
export async function copyUpstreamUrl(repo: RepoNode) {
    await env.clipboard.writeText(repo.parent!.html_url);
}

/**
 * Open the upstream URL in the browser
 *
 * @export
 * @param {RepoNode} node The repository to open the upstream URL from
 */
export function showUpstream(node: RepoNode) {
    let url = node.parent!.html_url;

    if (url) {
        env.openExternal(Uri.parse(url));
        output?.info(`Show upstream ${url}`);
    }
}

/**
 * Toggle visibility of a GitHub repository
 *
 * @export
 * @async
 * @param {RepoNode} repo The repository to toggle the visibility of
 * @returns {*}
 */
export async function toggleRepoVisibility(repo: RepoNode) {
    let visibility = repo.private ? "public" : "private";
    let confirm = await window.showWarningMessage(`Are you sure you want to make "${repo.name}" ${visibility}?`, { modal: true }, "Yes", "No");

    if (confirm !== "Yes") {
        return;
    }

    output?.info(`Toggle visibility of ${repo.full_name} to ${visibility}`);
    await updateGitHubRepository(repo, !repo.private, undefined);
    repo.private = !repo.private;
    store.init();
}
